### PR TITLE
Move compile fail tests

### PR DESCRIPTION
**Pull request by BD103**
# Objective

- Follow-up of #13184 :)
- We use `ui_test` to test compiler errors for our custom macros.
- There are four crates related to compile fail tests
  - `bevy_ecs_compile_fail_tests`, `bevy_macros_compile_fail_tests`, and `bevy_reflect_compile_fail_tests`, which actually test the macros.
  - [`bevy_compile_test_utils`](https://github.com/bevyengine/bevy/tree/64c1c65783938facc59d9b36cbaa6deba435d84e/crates/bevy_compile_test_utils), which provides helpers and common patterns for these tests.
- All of these crates reside within the `crates` directory.
- This can be confusing, especially for newcomers. All of the other folders in `crates` are actual published libraries, except for these 4.

## Solution

- Move all compile fail tests to a `compile_fail` folder under their corresponding crate.
  - E.g. `crates/bevy_ecs_compile_fail_tests` would be moved to `crates/bevy_ecs/compile_fail`.
- Move `bevy_compile_test_utils` to `tools/compile_fail_utils`.

There are a few benefits to this approach:

1. An internal testing detail is less intrusive (and confusing) for those who just want to browse the public Bevy interface.
2. Follows a pre-existing approach of organizing related crates inside a larger crate's folder.
   - See `bevy_gizmos/macros` for an example.
4. Makes consistent the terms `compile_test`, `compile_fail`, and `compile_fail_test` in code. It's all just `compile_fail` now, because we are specifically testing the error messages on compiler failures.
   - To be clear it can still be referred to by these terms in comments and speech, just the names of the crates and the CI command are now consistent.

## Testing

Run the compile fail CI command:

```shell
cargo run -p ci -- compile-fail
```

If it still passes, then my refactor was successful.

### CODE REVIEW

**Comment by BD103** [comment]
cc @Brezak, who previously worked on migrating our compile fail tests to `ui_test` in #12810.

**Comment by BD103** [comment]
The files changed (105) may be intimidating, but it's not actually that bad. The actual lines changed are just , the rest are just moved, unmodified files.

image

File: crates/bevy_reflect/compile_fail/README.md
Comment (Brezak) [inline crates/bevy_reflect/compile_fail/README.md:1]: This one will be broken too.

File: tools/compile_fail_utils/README.md
Comment (Brezak) [inline tools/compile_fail_utils/README.md:1]: The [tutorial](https://github.com/bevyengine/bevy/blob/311965300d7c5e80b8f8f3a22c11b6d8edae82f0/tools/compile_fail_utils/README.md?plain=1#L32-L38) on creating compile fail tests is outdated.

**Review by Brezak — CHANGES_REQUESTED** [review]
Looks ok. The documentation just needs to be fixed

**Review by alice-i-cecile — COMMENTED** [review]
I agree with the changes here: these are way less confusing if they're not top-level crates.

Once the links are fixed this LGTM.

**Review by alice-i-cecile — APPROVED** [review]

**Review by soqb — APPROVED** [review]
LGTM.

side note: i feel like we shouldn't put compile-succeed tests in these crates, since its much easier iteratively to drop them into regular `#[test]`s, since these tests are still flaky outside of ci.

**Review by Brezak — APPROVED** [review]

**Pull request merged**